// jspractice/index.js
var _ = require("lodash");

var ap = require("./array-problems");
exports.arrayProblems = ap.arrayProblems;

var cp = require("./conditional-problems");
exports.conditionalProblems = cp.conditionalProblems;

var fp = require("./function-problems");
exports.functionProblems = fp.functionProblems;

var sp = require("./string-problems");
exports.stringProblems = sp.stringProblems;

exports.testSolution = function(problemSet, problemIndex, body) {
    let problem = problemSet[problemIndex];

    console.log("Creating function...");
    let func = createFunction(body, problem.func.args);
    console.log(func);

    let completed = true;
    let returnHtml = "";

    problem.tests.forEach(function (item, index) {
        let result, correct;
        try {
            // save now, in case it crashes when trying
            // to run the function
            result = func(...item.args);
            correct = _.isEqual(result, item.ans);
        } catch (err) {
            result = err;
            correct = false;
        } finally {
            returnHtml += createTableRow(problem.func.name, item, result, correct);
            completed = correct && completed;
        }
    });

    return returnHtml;
}

function createFunction(body, functionArgs) {
    body = body.substring(body.indexOf("{") + 1, body.lastIndexOf("}")).trim();
    console.log("createFunction: ", body);
    return new Function(...functionArgs, body);
}

function createTableRow(funcName, test, result, correct) {
    return `<tr>
        <td>${funcName}(${quotedArgs(test.args)});</td>
        <td>${quotedString(test.ans)}</td>
        <td>${quotedString(result)}</td>
        <td>${correct ? "OK" : "X"}</td>
        <td class="${correct ? 'correct' : 'incorrect'}">&nbsp;</td>
     </tr>`;
}

function quotedString(ans) {
    if (typeof ans === "string")
        return `"${ans}"`;
    else if (ans instanceof Array)
        return `[ ${ans.join(", ")} ]`;
    return ans;
}

function quotedArgs(args) {
    let newArgs = [];
    args.forEach(function (item, index) {
        if (typeof item === "string")
            newArgs.push(`"${item}"`);
        else if (item instanceof Array)
            newArgs.push(`[ ${item.join(", ")} ]`);
        else
            newArgs.push(item);
    });
    return newArgs.join(", ");
}
